### PR TITLE
Refactoring of FunctionBlock

### PR DESCRIPTION
**Pull request by TheMSB**
Refactored FunctionBlock in order to make feature expansion easier.
FunctionBlock now extends Block and certain methods have subsequently been moved from FunctionBlock to Block.

newInstance of FunctionBlock has been slightly rewritten and a second version of newInstance has been introduced to enable the instantiation of FunctionBlocks with initialised names.

### CODE REVIEW

File: Code/src/main/java/nl/utwente/group10/ui/components/Block.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+public class Block extends StackPane {
+
+	/** Selected state of this Block*/
+	private boolean isSelected = false;
+	
+	/**
+	 * Set the selected boolean state of this Block
+	 * @param selectedState
+	 */
+		public void setSelected(boolean selectedState){
```
Comment (wandernauta) [inline Code/src/main/java/nl/utwente/group10/ui/components/Block.java]: Nitpick: this seems to be indented just one level further than necessary.

File: Code/src/main/java/nl/utwente/group10/ui/components/Block.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+			//TODO If another object is selected then deselect it first!!
```
Comment (wandernauta) [inline Code/src/main/java/nl/utwente/group10/ui/components/Block.java]: Is this the right place to do this? Would you expect foo.setSelected(true) to affect bar.isSelected() even if foo and bar appear to be unrelated?

File: Code/src/main/java/nl/utwente/group10/ui/components/FunctionBlock.java [hunk 3]
```diff
@@ -18,11 +18,11 @@
  * represents a Haskell function together with it's arguments and
  * visual representation.
  */
-public class FunctionBlock extends StackPane{
-	//The arguments this FunctionBlock holds
+public class FunctionBlock extends Block{
```
Comment (wandernauta) [inline Code/src/main/java/nl/utwente/group10/ui/components/FunctionBlock.java]: Missing space before {.

File: Code/src/main/java/nl/utwente/group10/ui/components/FunctionBlock.java [hunk 4]
```diff
@@ -31,32 +31,54 @@
...
-		//TODO Prettier resource loading
-		//TODO better factory
-		FunctionBlock functionBlock = null;
+		FunctionBlock functionBlock;
 		try{
 			functionBlock = (FunctionBlock) FXMLLoader.load(Main.class.getResource("/ui/FunctionBlock.fxml"), null, new TactileBuilderFactory());
 			functionBlock.initializeArguments(numberOfArguments);
 		}catch(IOException e){
 			e.printStackTrace();
+			functionBlock = null;
```
Comment (wandernauta) [inline Code/src/main/java/nl/utwente/group10/ui/components/FunctionBlock.java]: I don't like this catch. All this will do is cause NullPointerExceptions down the line, so why catch it at all?

File: Code/src/main/java/nl/utwente/group10/ui/components/FunctionBlock.java [hunk 5]
```diff
@@ -31,32 +31,54 @@
...
-	//TODO To be implemented method that will parse code into Haskell interface
-	// and returns the result
+	/**
+	 * Method that creates a newInstance of this class along with it's visual representation
+	 * @param the number of arguments this FunctionBlock can hold
+	 * @param the name of this FunctionBlock
+	 * @return a new instance of this class
+	 * @throws IOException
+	 */
+	public static FunctionBlock newInstance(int numberOfArguments, String name) throws IOException{
```
Comment (wandernauta) [inline Code/src/main/java/nl/utwente/group10/ui/components/FunctionBlock.java]: Missing space.

File: Code/src/main/java/nl/utwente/group10/ui/components/FunctionBlock.java [hunk 6]
```diff
@@ -31,32 +31,54 @@
...
 	public String executeMethod(){		
 		return new String("DEBUG-OUTPUT");
 	}
 	
+	/**
+	 * Nest another Node object within this FunctionBlock
+	 * @param node to nest
+	 */
 	public void nest(Node node){
 		Pane nestSPace = (Pane) this.lookup("#nest_space");
```
Comment (wandernauta) [inline Code/src/main/java/nl/utwente/group10/ui/components/FunctionBlock.java]: Is it a `nestSPace`, `nestSpace`, or `nestPane`?

File: Code/src/main/java/nl/utwente/group10/ui/components/FunctionBlock.java [hunk 7]
```diff
@@ -31,32 +31,54 @@
...
 		((Label) this.lookup("#label_function_name")).setText("Higher order function");
 		nestSPace.getChildren().add(node);
 	}
 	
-	//Private method to initialize the argument fields for this function block.
-	//All arguments will be stored as Strings.
+	/**
+	 * Private method to initialize the argument fields for this function block.
```
Comment (wandernauta) [inline Code/src/main/java/nl/utwente/group10/ui/components/FunctionBlock.java]: Nitpick: Documentation is not very useful.

File: Code/src/main/java/nl/utwente/group10/ui/components/FunctionBlock.java [hunk 8]
```diff
@@ -31,32 +31,54 @@
...
+	 * All arguments will be stored as Strings.
```
Comment (wandernauta) [inline Code/src/main/java/nl/utwente/group10/ui/components/FunctionBlock.java]: What would those Strings look like, exactly?

**Comment by wandernauta** [comment]
Big up for the refactoring, great idea. Some minor nitpicks and questions above, other than that :+1:

File: Code/src/main/java/nl/utwente/group10/ui/components/Block.java (hunk 2 above)
Comment (TheMSB) [inline Code/src/main/java/nl/utwente/group10/ui/components/Block.java]: Any Block within the UI can be selected, that may be extended beyond just FunctionBlocks. (think TypeBlock or OutputBlock), since they would all share this function I see no need to implement it multiple times in each respective extended class.

Comment (wandernauta) [inline Code/src/main/java/nl/utwente/group10/ui/components/Block.java]: Very true, but I was specifically referring to the TODO: would the setSelected method deselect another object, and how would it find that other object? Or should this be in the documentation, i.e. should someone else who calls this method call setSelected on the other object first?

File: Code/src/main/java/nl/utwente/group10/ui/components/FunctionBlock.java (hunk 4 above)
Comment (TheMSB) [inline Code/src/main/java/nl/utwente/group10/ui/components/FunctionBlock.java]: Catch was already in place, I just placed the null initialisation in a more appropriate place.
If so desired the try catch can be removed.

Comment (wandernauta) [inline Code/src/main/java/nl/utwente/group10/ui/components/FunctionBlock.java]: Yes please :relaxed:

**Comment by TheMSB** [comment]
Committed code styling fixes, will rework other code after discussion.

**Comment by wandernauta** [comment]
Sounds good!

File: Code/src/main/java/nl/utwente/group10/ui/components/FunctionBlock.java (hunk 4 above)
Comment (TheMSB) [inline Code/src/main/java/nl/utwente/group10/ui/components/FunctionBlock.java]: Fixed in next commit.

File: Code/src/main/java/nl/utwente/group10/ui/components/FunctionBlock.java (hunk 6 above)
Comment (TheMSB) [inline Code/src/main/java/nl/utwente/group10/ui/components/FunctionBlock.java]: Good question, as I did not develop this piece of code but just provided documentation from what I understood of it. I believe the name should be nestSpace and as such will be fixed in next commit.

File: Code/src/main/java/nl/utwente/group10/ui/components/FunctionBlock.java (hunk 8 above)
Comment (TheMSB) [inline Code/src/main/java/nl/utwente/group10/ui/components/FunctionBlock.java]: Strings will contain exactly what is entered though the user interface into the argument field, Since there is no way to actually set them as of yet but the need to store them is clear the initialisation has already been done. Next commit will contain a less obscure javaDoc.

File: Code/src/main/java/nl/utwente/group10/ui/components/Block.java (hunk 2 above)
Comment (TheMSB) [inline Code/src/main/java/nl/utwente/group10/ui/components/Block.java]: If an object is selected through the 'tap' gesture, other objects would have to be un-selected since the 'tap' gesture means: select THIS specific object. Hence the TODO. It is however still undetermined how group selections will be implemented exactly.

**Comment by TheMSB** [comment]
Commit second batch of changes, all issues *should* be resolved.

**Comment by wandernauta** [comment]
Looks good to me, will merge when Travis is green.

File: Code/src/main/java/nl/utwente/group10/ui/components/Block.java (hunk 2 above)
Comment (wandernauta) [inline Code/src/main/java/nl/utwente/group10/ui/components/Block.java]: Okay, sounds reasonable.

**Comment by wandernauta** [comment]
Merged, pleasure doing business! :shipit:
